smithy-model/src/main/resources/software/amazon/smithy/model/loader/prelude-traits.smithy

```diff
@@ -0,0 +1,516 @@
+       structurallyExclusive: true)
+@tags(["diff.error.const"])
+structure httpPayload {}
+
+/// "Defines an HTTP response code for an operation error."
+@trait(selector: "structure[trait|error]")
+@tags(["diff.error.const"])
+integer httpError
+
+/// Defines how a service supports cross-origin resource sharing.
+@trait(selector: "service")
+@tags(["diff.error.remove"])
+structure cors {
+    /// The origin from which browser script-originating requests will be allowed.
+    ///
+    /// Defaults to *.
+    origin: NonEmptyString,
+
+    /// The maximum number of seconds for which browsers are allowed to cache
+    /// the results of a preflight OPTIONS request.
+    ///
+    /// Defaults to 600, the maximum age permitted by several browsers.
+    /// Set to -1 to disable caching entirely.
+    maxAge: Integer,
+
+    /// The names of headers that should be included in the
+    /// Access-Control-Allow-Headers header in responses to preflight OPTIONS
+    /// requests. This list will be used in addition to the names of all
+    /// request headers bound to an input data member via the httpHeader, as
+    /// well as any headers required by the protocol or authentication scheme.
+    additionalAllowedHeaders: NonEmptyStringList,
+
+    /// The names of headers that should be included in the
+    /// Access-Control-Expose-Headers header in all responses sent by the
+    /// service. This list will be used in addition to the names of all
+    /// request headers bound to an output data member via the httpHeader,
+    /// as well as any headers required by the protocol or authentication
+    /// scheme.
+    additionalExposedHeaders: NonEmptyStringList,
+}
+
+@private
+list NonEmptyStringList {
+    member: NonEmptyString,
+}
+
+/// Marks a member as the payload of an event.
+@trait(selector: "member:of(structure):test(> :each(blob, string, structure, union))",
+       conflicts: [eventHeader],
+       structurallyExclusive: true)
+@tags(["diff.error.const"])
+structure eventPayload {}
+
+/// Marks a member as a header of an event.
+@trait(selector: "member:of(structure):test( > :each(boolean, byte, short, integer, long, blob, string, timestamp))",
+       conflicts: [eventPayload])
+@tags(["diff.error.const"])
+structure eventHeader {}
+
+/// Defines the operation input member name that contains an event stream.
+@trait(selector: "operation:test(-[input]->)")
+@tags(["diff.error.const"])
+string inputEventStream
+
+/// Defines the operation input member name that contains an event stream.
+@trait(selector: "operation:test(-[output]->)")
+@tags(["diff.error.const"])
+string outputEventStream
+
+@trait(selector: ":test(string, member > string)")
+structure idRef {
+    selector: String,
+    failWhenMissing: PrimitiveBoolean,
+    errorMessage: String,
+}
+
+@trait(selector: ":test(timestamp, member > timestamp)")
+@tags(["diff.error.const"])
+@enum("date-time": {},
+      "epoch-seconds": {},
+      "http-date": {})
+string timestampFormat
```

Comment:
This could use documentation since it's being migrated.

Reply:
done